cmd/gapic-showcase/compliance_suite_test.go

```diff
@@ -0,0 +1,203 @@
+
+	// Set handlers for each test case. When GAPIC generator tests do this, they should have
+	// each of their handlers invoking the correct GAPIC library method for the Showcase API.
+	restRPCs := map[string]prepRepeatDataTestFunc{
+		"Compliance.RepeatDataBody":       prepRepeatDataBodyTest,
+		"Compliance.RepeatDataQuery":      prepRepeatDataQueryTest,
+		"Compliance.RepeatDataSimplePath": prepRepeatDataSimplePathTest,
+	}
+
+	for _, group := range suite.GetGroup() {
+		rpcsToTest := group.GetRpcs()
+		for requestIdx, requestProto := range group.GetRequests() {
+			for rpcIdx, rpcName := range rpcsToTest {
+				errorPrefix := fmt.Sprintf("[request %d/%q: rpc %q/%d/%q]",
+					requestIdx, requestProto.GetName(), group.Name, rpcIdx, rpcName)
+
+				// Ensure that we issue only the RPCs the test suite is expecting.
+				rpcPrep, ok := restRPCs[rpcName]
+				if !ok {
+					t.Errorf("%s could not find prep function for this RPC", errorPrefix)
+					continue
+				}
+
+				prepName, verb, path, requestBody, err := rpcPrep(requestProto)
+				if err != nil {
+					t.Errorf("%s error: %s", errorPrefix, err)
+				}
+				if got, want := prepName, rpcName; got != want {
+					t.Errorf("%s retrieved mismatched prep function: got %q, want %q", errorPrefix, got, want)
+				}
+
+				// Issue the request. When GAPIC generator tests do this, they should simply
+				// invoke the correct GAPIC library method for the Showcase API.
+				httpRequest, err := http.NewRequest(verb, server.URL+path, strings.NewReader(requestBody))
+				if err != nil {
+					t.Errorf("%s error creating request: %s", errorPrefix, err)
+					continue
+				}
+				httpResponse, err := http.DefaultClient.Do(httpRequest)
+				if err != nil {
+					t.Errorf("%s error issuing call: %s", errorPrefix, err)
+					continue
+				}
+
+				// Check for successful response.
+				if got, want := httpResponse.StatusCode, 200; got != want {
+					t.Errorf("%s response code: got %d, want %d\n   %s %s",
+						errorPrefix, got, want, verb, server.URL+path)
+				}
+
+				// Unmarshal httpResponse body, interpreted as JSON.
+				// should do this.
+				responseBody, err := ioutil.ReadAll(httpResponse.Body)
+				httpResponse.Body.Close()
+				if err != nil {
+					t.Errorf("%s could not read httpResponse body: %s", errorPrefix, err)
+					continue
+				}
+				var response genproto.RepeatResponse
+				if err := jsonpb.UnmarshalString(string(responseBody), &response); err != nil {
+					t.Errorf("%s could not unmarshal httpResponse body: %s\n   response body: %s",
+						errorPrefix, err, string(responseBody))
+					continue
+				}
+
+				// Check for expected response.
+				if got, want := response.GetInfo(), requestProto.GetInfo(); !proto.Equal(got, want) {
+					gotString := jsonMarshaller.Format(got)
+					wantString := jsonMarshaller.Format(want)
+					t.Errorf("%s unexpected response:\n   -->got:\n`%s`\n   -->want:\n`%s`\n",
+						errorPrefix, gotString, wantString)
+				}
```

Comment:
I can do that, but is the output of `cmp.Diff` with `proto.Equal` humanly readable. What I like this (in general, not so useful for this test suite) is that if we very `got` is correct, we can paste it as the expected value

Reply:
Yeah the output is humanly readable. It produces (-/+) diffs of exactly what differed. This would be kind of bothersome to debug if broken because we'd have to "hand diff" the printed objects. I'd rather optimize for finding what is wrong. Up to you though, I'm not picky. :)

Updating "baseline" files or values is nice, but can be automated (we do this in gapic-generator-go via "update_golden" flag on the test suite) by having the test write the new value back to file.

Reply:
OK, let's try it.